Degree centrality implementation

Gohlub opened the pull request:
Proposed a solution for #1129. The implementation is straightforward, though I wonder whether I should add a flag to support in-degree and out-degree centrality, and support for weighted graphs (since the current implementation assumes an unweighted graph). Let me know what you think!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/rustworkx?pullRequest=1306) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/rustworkx?pullRequest=1306) before we can accept your contribution.<br/>**0** out of **3** committers have signed the CLA.<br/><br/>:x: ons<br/>:x: Gohlub<br/>:x: onsali<br/><hr/>**ons** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/rustworkx?pullRequest=1306) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/rustworkx?pullRequest=1306) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/rustworkx?pullRequest=1306) before we can accept your contribution.<br/>**0** out of **3** committers have signed the CLA.<br/><br/>:x: Gohlub<br/>:x: onsali<br/>:x: ons<br/><hr/>**ons** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/rustworkx?pullRequest=1306) it.</sub>

IvanIsCoding commented on .gitignore:
```diff
@@ -22,3 +22,5 @@ retworkx/*pyd
 **/*.so
 retworkx-core/Cargo.lock
+.DS_Store
+**/.DS_Store
```
A bit surprising that `.DS_Store` is not in there, but it should be fine. I think we only need the second instance though, right? It makes the first one redundant

IvanIsCoding commented on degree_centrality_implementation_rustworkx.md:
```diff
@@ -0,0 +1,149 @@
+
```
Do not include this with your pull request. I don't know if these were instructions for an AI assistant or a requirement for a college assignment, but it doesn't belong in the repository.

You should document the new method with a release note though.

IvanIsCoding commented on rustworkx-core/src/centrality.rs:
```diff
@@ -335,6 +335,48 @@ fn accumulate_edges<G>(
+    let mut centrality = vec![0.0; graph.node_bound()];
+
+    for node in graph.node_identifiers() {
+        let degree = graph.neighbors(node).count() as f64;
```
If you want to parametrize this to support `in_degree` and `out_degree`, call https://docs.rs/petgraph/latest/petgraph/graph/struct.Graph.html#method.neighbors_directed and accept an `Option<Directed>` https://docs.rs/petgraph/latest/petgraph/enum.Direction.html with `None, Some(), Some(Outgoing), Some(Incoming)` instead?

IvanIsCoding commented on rustworkx/__init__.pyi:
```diff
@@ -50,6 +50,7 @@ from .rustworkx import digraph_closeness_centrality as digraph_closeness_central
 from .rustworkx import graph_closeness_centrality as graph_closeness_centrality
 from .rustworkx import digraph_katz_centrality as digraph_katz_centrality
 from .rustworkx import graph_katz_centrality as graph_katz_centrality
+from .rustworkx import degree_centrality as degree_centrality
```
This will become `graph_degree_centrality` and `digraph_degree_centrality`

IvanIsCoding commented on src/centrality.rs:
```diff
@@ -165,6 +165,29 @@ pub fn digraph_betweenness_centrality(
+/// :rtype: CentralityMapping
+#[pyfunction]
+#[pyo3(text_signature = "(graph, /)")]
+pub fn degree_centrality_py(graph: &graph::PyGraph) -> PyResult<CentralityMapping> {
```
```suggestion
pub fn graph_degree_centrality(graph: &graph::PyGraph) -> PyResult<CentralityMapping> {
```
Stick to the existing naming convetion

IvanIsCoding commented on tests/graph/test_centrality.py:
```diff
@@ -230,3 +230,46 @@ def test_custom_graph_unnormalized(self):
+        self.graph.add_edges_from(edge_list)
+
+    def test_degree_centrality(self):
+        centrality = rustworkx.degree_centrality_py(self.graph)
```
```suggestion
        centrality = rustworkx.degree_centrality(self.graph)
```
A reminder to call the non-type specific function once you implement it in `__init__.py`

IvanIsCoding commented on tests/graph/test_centrality.py:
```diff
@@ -230,3 +230,46 @@ def test_custom_graph_unnormalized(self):
             self.assertAlmostEqual(v, expected[k])
+
+
+class TestDegreeCentrality(unittest.TestCase):
```
The existing test coverage is good, but remember to add a test with the multigraph case to check that we cover it properly

IvanIsCoding reviewed (commented):
Thanks for making your first contribution. I left some comments, but some things to note:
* Add release notes: https://github.com/Qiskit/rustworkx/blob/main/CONTRIBUTING.md#release-notes
* Add a universal function so your code works with both `PyGraph` and `PyDiGraph`. Here is a good example: https://github.com/Qiskit/rustworkx/blob/8774ccde2014e3a7e145e7bb14b4381b013f2d66/rustworkx/__init__.py#L1273
* You probably want to implement this centrality for `PyDiGraph` as well

Also, I noticed from the CLA bot that this PR has two authors. This is quite unusual for a first contribution so please let me know if you are:
1. Working on a college assignment together
2. One of the users flagged by the bot is an AI Assistant

There is nothing wrong with either, but it can help me review the PR accordingly.

IvanIsCoding commented:
Also, we can handle `in_degree` and `out_degree` in a separate PR. These methods would only be for `PyDiGraph`. They probably would reuse most of the existing code as I pointed out in the comment, but it requires more tests and all that.

Gohlub commented:
Hi, thank you for the swift feedback! I just wanted to note that this was in fact a college assignment, with the other contributor being my classmate. However we are now moving on to individual work, so I will be responsible for the next steps. I will be begin addressing your suggestions shortly.

IvanIsCoding commented:
> Hi, thank you for the swift feedback! I just wanted to note that this was in fact a college assignment, with the other contributor being my classmate. However we are now moving on to individual work, so I will be responsible for the next steps. I will be begin addressing your suggestions shortly.

Sounds good to me, just remember to ask your classmate(s) to sign the CLA otherwise the bot will not let it get merged

IvanIsCoding commented on rustworkx-core/src/centrality.rs:
````diff
@@ -335,6 +336,70 @@ fn accumulate_edges<G>(
+/// }
+/// ```
+
+pub fn graph_degree_centrality<G>(graph: G) -> Vec<f64>
````
For rustworkx-core, use a generic method that takes both kind of graphs. https://github.com/Qiskit/rustworkx/blob/aea56cf23b8a45716c1ae5af15a63b14222c8cf9/rustworkx-core/src/centrality.rs#L632 is a good example

IvanIsCoding commented on tests/graph/test_centrality.py:
```diff
@@ -230,3 +230,90 @@ def test_custom_graph_unnormalized(self):
+        }
+        self.assertEqual(expected, centrality)
+
+    def test_digraph_degree_centrality_with_direction(self):
```
Minor organization detail. This should put the digraph tests in tests/digraph/test_centrality.py with the other directed graph tests

Gohlub commented:
I tried to rebase to amend a commit my friend made, but I think I inadvertently added things to the PR that should not be there. What is the course of action here?

IvanIsCoding commented:
Keep "Maintainers are allowed to edit this pull request." active as it is right now and don't do further pushes. I will do a force push with the diff only and take it from here.

I will add you as an author and your classmate as co-author.

IvanIsCoding commented:
Just to let you know about the changes I made:
* I fixed some warnings from the lint (https://github.com/Qiskit/rustworkx/blob/main/CONTRIBUTING.md#lint)
* I added type annotations (https://github.com/Qiskit/rustworkx/blob/main/CONTRIBUTING.md#type-annotations)
* The rustdoc tests were broken so I had to fix them

Nevertheless, thanks for getting this started it was a lot of help
